File: ibus-switch/switch.c
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#define NMETHODS 2

int main() {
    static char* methods[NMETHODS];
    methods[0] = "xkb:us::eng";
    methods[1] = "pinyin";

    char buf[1024];
    // zero out buf
    memset(buf, 0, 1024);

    FILE* fp = popen("ibus engine", "r");
    int len = fread(buf, 1, sizeof(buf), fp);
    pclose(fp);

    if (len <= 0)
        return 0;

    if (buf[len-1] == '\n') {
        buf[len-1] = '\0';
    }
    for (int i = 0; i < NMETHODS; i++) {
        if (strcmp(buf, methods[i]) == 0) {
            sprintf(buf, "ibus engine %s", methods[(i+1) % NMETHODS]);
            system(buf);
            system("herbstclient emit_hook ibus");
            break;
        }
    }
}

File: ibus-switch/switch.go
package main

import "bufio"
import "os/exec"

var methods = [2]string{
	"xkb:us::eng",
	"pinyin",
}

func main() {
	cmd := exec.Command("ibus", "engine")
	out, _ := cmd.StdoutPipe()
	cmd.Start()
	r := bufio.NewScanner(out)
	if r.Scan() {
		currentMethod := r.Text()
		for i := 0; i < len(methods); i++ {
			// Try to find and switch
			if currentMethod == methods[i] {
				exec.Command("ibus", "engine", methods[(i+1)%len(methods)]).Run()
				exec.Command("herbstclient", "emit_hook", "ibus").Run()
				break
			}
		}
	}
}
